### PR TITLE
[New] create ignorePrivate option for no-multi-comp rule

### PR DESCRIPTION
**Pull request by vwillyams**
Adds a new option to no-multi-comp which ignores components that are not exported.

Currently WIP:
- Thorough testing, find edge cases. `npm test` wasn't working locally so I'm hoping to rely on CI for this.
- Comparisons between exports and React components probably needs more work. Maybe there's a pre-existing lib function for this. Maybe this functionality should be moved into lib.

### CODE REVIEW

File: docs/rules/no-multi-comp.md [hunk 1]
```diff
@@ -69,6 +69,55 @@ class HelloJohn extends React.Component {
 module.exports = HelloJohn;
 ```
 
+### `ignorePrivate`
```
Comment (ljharb) [inline docs/rules/no-multi-comp.md:72]: "private" might not be the best name here. maybe "ignoreNonExported"?

File: docs/rules/no-multi-comp.md [hunk 2]
```diff
@@ -69,6 +69,55 @@ class HelloJohn extends React.Component {
...
+```
+
+```jsx
+function Hello(props) {
+  return <div>Hello {props.name}</div>;
+}
+function HelloAgain(props) {
+  return <div>Hello again {props.name}</div>;
+}
+module.exports = {Hello, HelloAgain}
```
Comment (ljharb) [inline docs/rules/no-multi-comp.md:118]:
```suggestion
module.exports = { Hello, HelloAgain };
```

File: lib/rules/no-multi-comp.js [hunk 3]
```diff
@@ -46,6 +51,43 @@ module.exports = {
   create: Components.detect((context, components, utils) => {
     const configuration = context.options[0] || {};
     const ignoreStateless = configuration.ignoreStateless || false;
+    const ignorePrivate = configuration.ignorePrivate || false;
+
+    const exportedComponents = new Set(); // Track exported components
+    const validIdentifiers = new Set(['ArrowFunctionExpression', 'Identifier', 'FunctionExpression']);
```
Comment (ljharb) [inline lib/rules/no-multi-comp.js:57]: i believe in react 19, async functions can be components as well.

File: lib/rules/no-multi-comp.js [hunk 4]
```diff
@@ -46,6 +51,43 @@ module.exports = {
...
+    function getExportedComponentName(node) {
+      if (node.declaration.type === 'ClassDeclaration') {
+        return node.declaration.id.name;
+      }
+      for (const declarator of node.declaration.declarations || []) {
+        const type = declarator.init.type;
+        if (validIdentifiers.has(type)) {
+          return declarator.id.name;
+        }
+      }
```
Comment (ljharb) [inline lib/rules/no-multi-comp.js:73]:
```suggestion
      if (node.declaration.declarations) {
        const declarator = node.declaration.declarations.find((declarator) => validIdentifiers.has(declarator.init.type));
        if (declarator) {
          return declarator.id.name;
        }
      }
```

File: lib/rules/no-multi-comp.js [hunk 5]
```diff
@@ -61,21 +103,46 @@ module.exports = {
...
+
+    const rule = {
       'Program:exit'() {
         if (components.length() <= 1) {
           return;
         }
 
         values(components.list())
           .filter((component) => !isIgnored(component))
+          .filter((component) => !isPrivate(component))
```
Comment (ljharb) [inline lib/rules/no-multi-comp.js:123]:
```suggestion
          .filter((component) => !isIgnored(component) && !isPrivate(component))
```

File: lib/rules/no-multi-comp.js [hunk 6]
```diff
@@ -61,21 +103,46 @@ module.exports = {
...
+
+    if (ignorePrivate) {
+      rule.ExportNamedDeclaration = (node) => {
+        exportedComponents.add(getExportedComponentName(node));
+      };
+
+      rule.ExportDefaultDeclaration = (node) => {
+        exportedComponents.add(getExportedComponentName(node));
+      };
+    }
```
Comment (ljharb) [inline lib/rules/no-multi-comp.js:144]: instead of mutating maybe we could use Object.assign here?

File: lib/rules/no-multi-comp.js (hunk 3 above)
Comment (vwillyams) [reply lib/rules/no-multi-comp.js:57]: Yeah good point, I had not thought about RSCs.

File: lib/rules/no-multi-comp.js (hunk 6 above)
Comment (vwillyams) [reply lib/rules/no-multi-comp.js:144]: Yeah sure, sounds good. I'm not clear on why but I'm not super experienced in the eslint lifecycle and I can imagine that mutation may introduce issues somehow.

File: docs/rules/no-multi-comp.md (hunk 1 above)
Comment (vwillyams) [reply docs/rules/no-multi-comp.md:72]: `ignoreNonExported` feels a little clunky to me but I can't think of anything better. just one of those hard problems in software.

Comment (ljharb) [reply docs/rules/no-multi-comp.md:72]: yeah totally :-) "ignoreInternal"?

File: lib/rules/no-multi-comp.js (hunk 6 above)
Comment (ljharb) [reply lib/rules/no-multi-comp.js:144]: it probably doesn't matter much at all, but mutation generally risks slowdowns later, as opposed to creating objects all at once.

File: lib/rules/no-multi-comp.js [hunk 7]
```diff
@@ -61,21 +103,46 @@ module.exports = {
...
+
+    const rule = {
       'Program:exit'() {
         if (components.length() <= 1) {
           return;
         }
 
         values(components.list())
           .filter((component) => !isIgnored(component))
```
Comment (ljharb) [inline lib/rules/no-multi-comp.js:122]:
```suggestion
```

Comment (vwillyams) [reply lib/rules/no-multi-comp.js:122]: ah yeah, the commit suggestion button resulted in some unhelpful commits. I'll resolve these manually.

File: docs/rules/no-multi-comp.md (hunk 1 above)
Comment (vwillyams) [reply docs/rules/no-multi-comp.md]: Sounds good to me, switched over.

File: lib/rules/no-multi-comp.js (hunk 3 above)
Comment (vwillyams) [reply lib/rules/no-multi-comp.js:57]: Added a little coverage for this - I'm not an RSC dev by day though so may have made some oversights.

File: lib/rules/no-multi-comp.js (hunk 6 above)
Comment (vwillyams) [reply lib/rules/no-multi-comp.js]: Switched over.

**Comment by vwillyams** [comment]
FYI I'm currently iterating on the test coverage and making sure this works in all cases. I didn't entirely expect the amount of complexity that would go into this feature! But it's been a fun learning process.

File: tests/lib/rules/no-multi-comp.js [hunk 8]
```diff
@@ -22,6 +22,178 @@ const parserOptions = {
...
+  (compOne, compOneName, compTwo, compTwoName) => `
+    export default ${compOne(compOneName)}
+    export ${compTwo(compTwoName)}`, // default export
+];
+
+/**
+ * @param {ComponentExportGenerator[]} scenarioArray array of scenario generator functions which we will now convert into strings
+ * @param {ComponentGenerator[]} componentTypes array of components to generate on
+ * @param {boolean} [invalid] whether generated scenarios should expect to fail
+ * @returns {{code: string, options: object[], errors: object[]}[]}
```
Comment (ljharb) [inline tests/lib/rules/no-multi-comp.js]: **Syntax error**: `module.exports = { ${compOneName} ${compTwoName} }` is missing a comma between the property shorthand names. Should be:
```js
module.exports = { ${compOneName}, ${compTwoName} }
```
This causes a parse error at runtime (3 test failures).

File: tests/lib/rules/no-multi-comp.js [hunk 9]
```diff
@@ -22,6 +22,178 @@ const parserOptions = {
   },
 };
 
+// ------------------------------------------------------------------------------
+// Combinatorial test generation for ignoreInvalid
+// ------------------------------------------------------------------------------
+// eslint-disable-next-line valid-jsdoc
+/**
+ * @typedef {Function} ComponentGenerator
```
Comment (ljharb) [inline tests/lib/rules/no-multi-comp.js]: The combinatorial test generation approach generates hundreds of tests, many of which pass or fail for the wrong reasons (because the rule doesn't detect the export form being tested). I'd suggest replacing this with simple, explicit test cases — like every other rule in this repo — so each test clearly documents what it's testing and can be individually debugged. The current approach makes it very hard to tell which combinations actually work vs. which silently pass by accident.

**Review by ljharb — CHANGES_REQUESTED** [review]
I ran the tests locally — **771 out of 1545 tests fail**. The issues fall into a few categories:

1. **Crash: `TypeError: Cannot read properties of null (reading 'type')`** (105 failures) — `export { Foo }` has `node.declaration === null`, so `getExportedComponentName` crashes on `node.declaration.type`.
2. **Crash: `TypeError: Cannot read properties of null (reading 'name')`** (108 failures) — `export default function() {}` has `node.declaration.id === null`, so `getExportedComponentName` crashes on `node.declaration.id.name`. Similarly, `findComponentIdentifierFromComponent` can crash when a component has no `id` (e.g. anonymous arrow or memo-wrapped functions).
3. **Silent failures: 555 invalid tests produce no errors** — many export forms are not detected, so components that should be "exported" are treated as private and silently ignored.
4. **Parse error** (3 failures) — `module.exports = { ComponentOne() {} ComponentTwo() {} }` is missing a comma between the methods.

Detailed issues are in the inline comments below. The core problem is that `getExportedComponentName` only handles a few specific AST shapes and silently returns `undefined` for the rest, which causes false negatives instead of visible failures.

**Review by ljharb — CHANGES_REQUESTED** [review]
added some review comments
